### PR TITLE
feat: enhance networking to support per replica service

### PR DESCRIPTION
**Pull request by JasonHe-WQ**
### Ⅰ. Motivation
#275 

### Ⅱ. Modifications

Align with `LeaderWorkerSet` and support better DNS discovery

### Ⅲ. Does this pull request fix one issue?

fixes #XXXX

## Checklist

- [ ] Format your code `make fmt`.
- [ ] Add unit tests or integration tests.
- [ ] Update the documentation related to the change.

### CODE REVIEW

File: keps/275-enhance-network/README.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+## Summary
+
+This KEP proposes a new role-level networking field,`NetworkConfig.SubdomainPolicy`, with two supported policies:
+
+- `Shared`
+- `UniquePerReplica`
+
+
+RBG currently creates one headless Service per role. That works for simple cases, but it is noy enough for some large scale cases.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:95]: Typo: "noy" → "not".

File: keps/275-enhance-network/README.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+In particular, `RoleInstanceSet + leaderWorkerPattern` has two practical needs that are not supported by the current controller:
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:97]: This sentence references `RoleInstanceSet + leaderWorkerPattern`, but Goal #3 (line 124) says `UniquePerReplica` is for `StatefulPattern`, and line 258 says "stateful `leaderWorkerPattern`". These three statements point to different pattern combinations. Please converge to a single, precise definition of which workload kind × role pattern × instance pattern supports `UniquePerReplica`, ideally in a table format.

File: keps/275-enhance-network/README.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+1. We would like to have segregated service for each replica to avert using native round robin load balance algorithms to enhance serving.
+2. At larger scale, a single shared headless Service can become a bottleneck to aggregate every Pod of the role. 
+Setting one headless Service per replica as `leaderworkerset` does so that each replica has its own DNS domain.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:101]: Typo/casing: `leaderworkerset` should be `LeaderWorkerSet` to match the API kind.

File: keps/275-enhance-network/README.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+More importantly, direct Pod DNS access through headless Services depends on the Pod `subdomain` matching the governing Service name.
+This means `UniquePerReplica` cannot be handled by extra Services only. It changes the actual network identity used by the Pods,
+so this KEP must define rollout behavior and compatibility carefully.
+
+### Goals
+
+1. Introduce a clear API for selecting the headless Service topology of a role while keeping current behavior as the default.
+2. Keep a shared Service for `RoleInstanceSet + leaderWorkerPattern`.
+3. Support one headless Service per `RoleInstance` in the same `Roleinstanceset` when using `StatefulPattern`.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:124]: Typo: `Roleinstanceset` should be `RoleInstanceSet`. Also, this Goal says `UniquePerReplica` is for `StatefulPattern`, but line 258 says "stateful `leaderWorkerPattern`" — please resolve this inconsistency.

File: keps/275-enhance-network/README.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+
+1. Supporting `standalonePattern` or `customComponentsPattern`.
+2. Supporting `Deployment`, `StatefulSet`, `LeaderWorkerSet`, or other role pattern.
+3. Introducing `LeaderOnly`  service in this KEP.
+4. Keeping every policy transition non-rolling. Transitions which change Pod DNS identity must rollout.
+
+## Proposal
+
+Add a new optional networking field in `RoleSpec` which contains there types:
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:137]: Typo: "there types" should be "two policies". Also, the `UniquePerReplica` bullet has grammar issues — consider: "creates a separate headless Service per replica, selecting only Pods in that replica, while each replica gets its own subdomain."

File: keps/275-enhance-network/README.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+#### Story 1
+
+When using `sglang` or `vllm` to serve large models, I want the requests target to the certain service instead of random one,
+and the algorithm may be `cache aware` or `least running length`, which is not supported through k8s native kube-proxy service.
+
+#### Story 2
+
+I have a large cluster. The replica is so large that the `CoreDNS` may fail to solve. I want each replica to have its own headless Service 
+so that DNS and endpoints can be looked up safely instead of be choped by connections.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:154]: Typo: "choped" → "chopped" (and "be chopped" → "being chopped" reads more naturally).

File: keps/275-enhance-network/README.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+#### Story 3
+I was using `LeaderWorkerSet` and I want to migrate to `RoleBasedGroup` using stateful pattern of `RoleInstanceSet` workload while not change any API field, 
+and I prefer smooth migration to `RoleBaseGroup` and not change API.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:158]: Typo: `RoleBaseGroup` → `RoleBasedGroup`.

File: keps/275-enhance-network/README.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+type RoleSpec struct {
+    // +optional
+    NetworkConfig *NetworkConfig `json:"networkConfig,omitempty"`
+}
+
+type NetworkConfig struct {
+    // SubdomainPolicy determines the policy that will be used when creating
+    // the headless service, defaults to shared
+    // +kubebuilder:validation:Enum={Shared,UniquePerReplica}
+    SubdomainPolicy *SubdomainPolicy `json:"subdomainPolicy"`
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:176]: The struct tag shows `json:"subdomainPolicy"` without `omitempty`, and lacks `+optional` and `+kubebuilder:default=Shared`. The KEP states the default is `Shared` when the field is empty, but the API snippet does not reflect this. Suggested fix:
```go
// +optional
// +kubebuilder:default=Shared
// +kubebuilder:validation:Enum={Shared,UniquePerReplica}
SubdomainPolicy *SubdomainPolicy `json:"subdomainPolicy,omitempty"`
```

File: keps/275-enhance-network/README.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+type SubdomainPolicy string
+
+const (
+    // replica 0: my-rbg-0-0.my-rbg, my-rbg-0-1.my-rbg
+    // replica 1: my-rbg-1-0.my-rbg, my-rbg-1-1.my-rbg
+    SubdomainShared SubdomainPolicy = "Shared"
+
+
+    // replica 0: my-rbg-0-0.my-rbg-0, my-rbg-0-0.my-rbg-0
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:187]: The `UniquePerReplica` DNS example has a copy/paste error: both entries show `my-rbg-0-0.my-rbg-0`. The second pod should be `my-rbg-0-1.my-rbg-0`:
```
// replica 0: my-rbg-0-0.my-rbg-0, my-rbg-0-1.my-rbg-0
```

File: keps/275-enhance-network/README.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+
+
+## Summary
+
+This KEP proposes a new role-level networking field,`NetworkConfig.SubdomainPolicy`, with two supported policies:
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:89]: The field is named `SubdomainPolicy`, but its actual scope covers service topology — service count, ownership, selector granularity, discovery generation, and rollout behavior — not just pod subdomain. Consider renaming to `ServiceDiscoveryPolicy`, `HeadlessServicePolicy`, or `ServiceTopologyPolicy` so the name aligns with what it governs. If you prefer to keep `SubdomainPolicy`, please add an explicit rationale section explaining why this field also defines service topology.

File: keps/275-enhance-network/README.md [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+### Behavior by Policy
+
+#### `Shared`
+
+This is current behavior and remains the default.
+- One shared headless Service for one role and includes all role Pods.
+- Existing workloads require no spec changes and see no behavioral change.
+
+#### `UniquePerReplica`
+- One unique headless Service for Pod in each replica within the same role.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:206]: Under `UniquePerReplica`, does the role-level shared headless Service still exist in steady state, or is it removed after rollout completes? This is a critical design decision that affects: (1) backward compatibility for discovery logic, (2) whether users can still use the role-level service name, (3) env var / ConfigMap output semantics. Please state the intended behavior explicitly and provide rationale.

File: keps/275-enhance-network/README.md [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+`Shared -> UniquePerReplica`  
+`UniquePerReplica -> Shared`  
+
+
+These require a `RoleInstanceSet` rollout. Transition rules:
+
+1. The controller first creates the target Services for the new policy.
+2. Pods are recreated through the normal rollout path so that their `subdomain` matches the new governing Service name.
+3. Old and new Services may temporarily coexist during rollout. And once the rollout is complete, old Services are deleted.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:235]: The rollout description is directionally correct but lacks controller-level convergence conditions. Please specify:
1. When is an old Service safe to delete? (e.g., all pods reference the new governing service, all RoleInstances ready?)
2. What happens on rollout failure — do both old and new Services persist? Is partial convergence allowed?
3. Please state explicitly that in-place DNS identity mutation is not supported; any policy transition must recreate pods.

File: keps/275-enhance-network/README.md [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+
+
+### Discovery and Environment Variables
+
+Current discovery and environment injection assume one role-level service name.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:240]: The discovery section correctly identifies that `UniquePerReplica` changes service-based addresses, but the contract is underspecified for implementation. Please define:
1. Which discovery artifacts are currently role-scoped? List them explicitly.
2. Under `UniquePerReplica`, how does each artifact change — single value → per-replica value? Return local replica or all replicas?
3. Do leader and worker pods get different discovery addresses?
4. How should existing consumers that assume "one role = one service name" migrate?

File: keps/275-enhance-network/README.md [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+### Service Naming and Owner Reference
+
+#### `Shared`
+- Service name continues to use the existing shared naming convention.
+- The shared Service is owned by the `RoleInstanceSet`.
+- transitions between `Shared` and `UniquePerReplica` change the governing service name of Pods
+
+#### `UniquePerReplica`
+- Each per-instance Service name is equal to the `RoleInstance` name.
+- The per-instance Service is owned by that `RoleInstance`.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:218]: Per-instance Service owned by `RoleInstance` is reasonable, but please add rationale for this ownership choice. Specifically: why not own all services under `RoleInstanceSet` and use labels/annotations for instance association? What happens when `RoleInstance` lifecycle is volatile — is service GC timing-sensitive? A one-line rationale would address reviewer concerns.

File: keps/275-enhance-network/README.md [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+
+Under `UniquePerReplica`, discovery becomes replica-scoped. Each `RoleInstance` uses its own governing Service name, so direct Pod DNS names change to:
+
+`<pod-name>.<role-instance-name>`
+
+Because of this, `UniquePerReplica` is not just an additive Service creation policy. It changes the service name used by Pods, 
+so discovery artifacts and environment variables that embed service addresses must be generated according to the active `SubdomainPolicy`.
+
+
+Only `RoleInstanceSet` with stateful `leaderWorkerPattern` can use policy field.
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:258]: This line says "Only `RoleInstanceSet` with stateful `leaderWorkerPattern` can use policy field", but Goal #3 says `UniquePerReplica` is for `StatefulPattern` only. These are different pattern types. Please resolve this and provide a clear supported-matrix table specifying: workload kind × role pattern × instance pattern × which policies are allowed, and what happens for unsupported combinations (admission reject vs controller ignore).

File: keps/275-enhance-network/README.md [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+
+[ ] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes
+necessary to implement this enhancement.
+
+
+##### Unit tests
+
+- API defaulting for `Shared`
+- Per replica Service name for `UniquePerReplica`
```
Comment (cheyang) [inline keps/275-enhance-network/README.md:271]: The test plan covers the basics but misses important edge cases:
1. Existing workloads without the field see zero behavioral change
2. Unsupported pattern + `UniquePerReplica` is rejected
3. Discovery/env compatibility across policies
4. Mid-rollout failure — are objects consistent?
5. Old and new Services coexisting — DNS/selector does not cross-contaminate
6. Scale up/down — per-replica Service create/delete correctness
7. Owner deletion / GC behavior for per-instance services

**Review by cheyang — CHANGES_REQUESTED** [review]
I think the KEP needs clearer definitions on supported combinations, steady-state resource model, discovery contract, and transition mechanics before proceeding to implementation.

Key areas to address:
1. The supported matrix is inconsistent across Goals, Proposal, and Validation sections.
2. It is unclear whether a role-level shared Service is retained in steady-state under `UniquePerReplica`.
3. The field name `SubdomainPolicy` models service topology, not just subdomain behavior — consider renaming or documenting the rationale.
4. The discovery / env contract is underspecified.
5. Rollout transition conditions and failure handling need more detail.

Detailed comments are inline below.

**Review by Syspretor — APPROVED** [review]
/lgtm

**Review by cheyang — APPROVED** [review]
